src/libcore/fmt/mod.rs

```diff
@@ -272,6 +270,14 @@ impl<'a> ArgumentV1<'a> {
     #[doc(hidden)]
     #[unstable(feature = "fmt_internals", reason = "internal to format_args!", issue = "none")]
     pub fn new<'b, T>(x: &'b T, f: fn(&T, &mut Formatter<'_>) -> Result) -> ArgumentV1<'b> {
+        // SAFETY: `mem::transmute(x)` is safe because
+        //     1. `&'b T` is transmuted to have an explicit lifetime `'b`
+        //              (so as to not have an unbounded lifetime)
+        //     2. `&'b T` and `&'b Void` have the same memory layout
+        //              (as long as `T` is `Sized`)
+        // `mem::transmute(f)` is safe since `fn(&T, &mut Formatter<'_>) -> Result`
+        // and `fn(&Void, &mut Formatter<'_>) -> Result` have the same ABI
+        // (as long as `T` is `Sized`)
```

Comment:
I am not 100% sure this is true since Void is a ZST, though I guess it does work out today since this code works. I would like to note here if this is not a detail that we guarantee to work, though. cc @eddyb

Reply:
Oh god we haven't fixed that yet?
`enum Void {}` should be replaced with an `extern { type Erased; }`.

Reply:
It's not a uninhabited type -- it's a ZST. Not sure exactly what we haven't fixed though. I agree that the extern type is the better solution here -- we should leave a FIXME here.

```rust
struct Void {
    _priv: (),
    /// Erases all oibits, because `Void` erases the type of the object that
    /// will be used to produce formatted output. Since we do not know what
    /// oibits the real types have (and they can have any or none), we need to
    /// take the most conservative approach and forbid all oibits.
    ///
    /// It was added after #45197 showed that one could share a `!Sync`
    /// object across threads by passing it into `format_args!`.
    _oibit_remover: PhantomData<*mut dyn Fn()>,
}
```

Reply:
> It's not a uninhabited type -- it's a ZST.

Okay so we fixed that, not as scary as it used to be. The name is *very* wrong though, it should be `Opaque` or something similar. `Void` should be reserved for empty enums.

Reply:
I added a FIXME for `Void`, but I wasn't sure if what I said about `mem::transmute(f)` being safe needed to be changed. (Is that true?)

Reply:
mem::transmute is fine, resolving.